nncf/pruning/filter_pruning/global_ranking/RL_evolution.py

```diff
@@ -0,0 +1,212 @@
+                action[k] = (scale, shift)
+        else:
+            self.mean_reward.append(np.mean(self.population_reward))
+            sampled_idxs = np.random.choice(self.POPULATIONS, self.SAMPLES)
+            sampled_rewards = self.population_reward[sampled_idxs]
+            winner_idx_ = np.argmax(sampled_rewards)
+            winner_idx = sampled_idxs[winner_idx_]
+
+            # Mutate winner
+            winner = self.population_data[winner_idx]
+            # Perturbate winner
+            mutate_num = int(self.mutate_percent * self.num_layers)
+            mutate_candidate_idxs = np.random.choice(self.num_layers, mutate_num)
+            mutate_candidates = self.all_action_keys[mutate_candidate_idxs]
+            for k in self.all_action_keys:
+                scale = 1
+                shift = 0
+                if k in mutate_candidates:
+                    scale = np.exp(float(np.random.normal(0, self.SCALE_SIGMA * step_size)))
+                    shift = float(np.random.normal(0, self.original_dist_stat[k]['std']))
+                action[k] = (scale * winner[k][0], shift + winner[k][1])
+            self.oldest_index = self.index_queue.get()
+        return action
+
+    def ask(self, episode_num):
+        """
+        Predict and returns action for the last told episode information: state, reward, episode_num and info
+        :return:
+        """
+        self.episode = episode_num
+        action = self._predict_action()
+        self.last_action = action
+        return action
+
+    def tell(self, state, reward, end_of_episode, episode_num, info):
+        """
+        Getting info about episode step and save it every end of episode
+        """
+        # save state, reward and info from the current step
+        self.state = state
+        self.reward = reward
+        self.episode = episode_num
+        self.info = info
+
+        if end_of_episode:
+            self._save_episode_info(reward)
+
+
+class LeGREvolutionEnv():
+    def __init__(self, filter_pruner, model, train_loader, val_loader, train_fn, val_fn, config,
+                 train_steps, pruning_max):
+        self.loss_as_reward = True
+        self.prune_target = pruning_max
+
+        # Train/test params
+        self.train_loader, self.val_loader = train_loader, val_loader
+        self.train_fn = train_fn
+        self.validate_fn = val_fn
+        self.config = config
+
+        self.filter_pruner = filter_pruner
+        self.steps = train_steps
+
+        self.model = model
+
+    def reset(self):
+        self.filter_pruner.reset()
+        self.model.eval()
+
+        self.full_flops = self.filter_pruner.get_flops_number_in_model()
+        self.rest = self.full_flops
+        self.last_act = (1, 0)
+
+        return torch.zeros(1), [self.full_flops, self.rest]
+
+    def train_steps(self, steps):
+        optimizer = optim.SGD(self.model.parameters(), lr=1e-2, momentum=0.9, weight_decay=5e-4, nesterov=True)
```

Comment:
Ideally we would let the user register the optimizer along with other init args, not hardcode it here. SGD w/o learning rate warmup could lead to unstable results during initial fine-tuning. We could probably try out a more robust alternative, like RAdam.

Reply:
Added train_optimizer for LeGR as init argument, this optimizer left as default option.